src/main/java/org/spongepowered/api/data/type/RailDirection.java

```diff
@@ -24,11 +24,103 @@
  */
 package org.spongepowered.api.data.type;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.Sponge;
 import org.spongepowered.api.util.Cycleable;
+import org.spongepowered.api.util.Direction;
 import org.spongepowered.api.util.annotation.CatalogedBy;
 
+import java.util.Optional;
+
 @CatalogedBy(RailDirections.class)
 public interface RailDirection extends CatalogType, Cycleable<RailDirection> {
 
+    /**
+     * Returns the {@link RailDirection} from the given facing directions.
+     *
+     * @param firstDirection The first direction
+     * @param secondDirection The second direction
+     * @return The {@link RailDirection}
+     * @throws IllegalArgumentException If the given directions don't differ
+     */
+    static RailDirection fromFaces(final Direction firstDirection, final Direction secondDirection) {
+        checkNotNull(firstDirection, "firstDirection");
+        checkNotNull(secondDirection, "secondDirection");
+
+        Direction cardinalFirstDirection = Direction.getClosest(firstDirection.asOffset(), Direction.Division.CARDINAL);
+        Direction cardinalSecondDirection = Direction.getClosest(secondDirection.asOffset(), Direction.Division.CARDINAL);
+
+        if (cardinalFirstDirection.equals(cardinalSecondDirection)) {
+            throw new IllegalArgumentException("The directions should be different!");
+        }
+
+        for (RailDirection currentRailDirection: Sponge.getRegistry().getAllForMinecraft(RailDirection.class)) {
+            if (currentRailDirection.isFacing(cardinalFirstDirection) && currentRailDirection.isFacing(cardinalSecondDirection)) {
+                return currentRailDirection;
+            }
+        }
+
+        throw new AssertionError();
+    }
+
+    /**
+     * Returns the {@link RailDirection} from the given ascending direction.
+     *
+     * @param ascendingDirection The ascending direction
+     * @return The {@link RailDirection}
+     */
+    static RailDirection fromAscendingDirection(final Direction ascendingDirection) {
```

Comment:
This feels very confusing to me - and the javadocs aren't particularly informative as to why there is one direction, not two.

`fromAscendingDirection` doesn't really make any sense from a language point of view. I know what you're trying to say, that you are defining the direction where the rail terminus is at the top of the block, not the bottom, and that by Minecraft logic it must be the opposite direction which is the "direction" of the bottom terminus.

`withUpperDirection`/`withUpperAxis`?

Opinions?